File: Heap/MedianHeap.java
import java.util.Collections;
import java.util.PriorityQueue;

public class MedianHeap {
	PriorityQueue<Integer> minHeap;
	PriorityQueue<Integer> maxHeap;

	public MedianHeap() {
		minHeap = new PriorityQueue<Integer>();
		maxHeap = new PriorityQueue<Integer>(Collections.reverseOrder());
	}

	// Other Methods.
	public void add(int value) {
		if (maxHeap.size() == 0 || maxHeap.peek() >= value) {
			maxHeap.add(value);
		} else {
			minHeap.add(value);
		}

		// size balancing
		if (maxHeap.size() > minHeap.size() + 1) {
			value = maxHeap.remove();
			minHeap.add(value);
		}

		if (minHeap.size() > maxHeap.size() + 1) {
			value = minHeap.remove();
			maxHeap.add(value);
		}
	}

	public int getMedian() {
		if (maxHeap.size() == 0 && minHeap.size() == 0)
			return Integer.MAX_VALUE;

		if (maxHeap.size() == minHeap.size())
			return (maxHeap.peek() + minHeap.peek()) / 2;
		else if (maxHeap.size() > minHeap.size())
			return maxHeap.peek();
		else
			return minHeap.peek();
	}

	public static void main(String[] args) {
		int arr[] = { 1, 9, 2, 8, 3, 7 };
		MedianHeap hp = new MedianHeap();

		for (int i = 0; i < 6; i++) {
			hp.add(arr[i]);
			System.out.println("Median after addition of " + arr[i] + " is  " + hp.getMedian());
		}
	}
}

/*
 * Median after addition of 1 is 1 
 * Median after addition of 9 is 5 
 * Median after addition of 2 is 2 
 * Median after addition of 8 is 5 
 * Median after addition of 3 is 3 
 * Median after addition of 7 is 5
 */

File: Heap/MedianHeap.js
var MedianHeap = /** @class */ (function () {
    function MedianHeap() {
        if (this.minHeap === undefined)
            this.minHeap = null;
        if (this.maxHeap === undefined)
            this.maxHeap = null;
        this.minHeap = (new java.util.PriorityQueue());
        this.maxHeap = (new java.util.PriorityQueue((java.util.Collections.reverseOrder())));
    }
    MedianHeap.prototype.insert = function (value) {
        if (this.maxHeap.size() === 0 || this.maxHeap.peek() >= value) {
            this.maxHeap.add(value);
        }
        else {
            this.minHeap.add(value);
        }
        if (this.maxHeap.size() > this.minHeap.size() + 1) {
            value = this.maxHeap.remove();
            this.minHeap.add(value);
        }
        if (this.minHeap.size() > this.maxHeap.size() + 1) {
            value = this.minHeap.remove();
            this.maxHeap.add(value);
        }
    };
    MedianHeap.prototype.getMedian = function () {
        if (this.maxHeap.size() === 0 && this.minHeap.size() === 0)
            return 2147483647;
        if (this.maxHeap.size() === this.minHeap.size())
            return ((this.maxHeap.peek() + this.minHeap.peek()) / 2 | 0);
        else if (this.maxHeap.size() > this.minHeap.size())
            return this.maxHeap.peek();
        else
            return this.minHeap.peek();
    };
    MedianHeap.main = function (args) {
        var arr = [1, 9, 2, 8, 3, 7, 4, 6, 5, 1, 9, 2, 8, 3, 7, 4, 6, 5, 10, 10];
        var hp = new MedianHeap();
        for (var i = 0; i < 20; i++) {
            {
                hp.insert(arr[i]);
                console.info("Median after insertion of " + arr[i] + " is  " + hp.getMedian());
            }
            ;
        }
    };
    return MedianHeap;
}());
MedianHeap["__class"] = "MedianHeap";
MedianHeap.main(null);
